dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -278,48 +284,58 @@ class LocalTunnelSender : public TunnelSender
         // is responsible for deleting receiver_mem_tracker must be destroyed after these local tunnels.
         data->switchMemTracker(local_request_handler.recv_mem_tracker);
 
-        auto res = local_request_handler.write<enable_fine_grained_shuffle>(source_index, data);
-
-        if (unlikely(!res))
-            closeLocalTunnel(true, "Push mpp packet failed at local tunnel");
-        return res;
+        return local_request_handler.write<enable_fine_grained_shuffle>(source_index, data);
     }
 
     void cancelWith(const String & reason) override
     {
-        closeLocalTunnel(true, reason);
+        prepareToCloseLocalTunnel(true, reason);
     }
 
     bool finish() override
     {
-        closeLocalTunnel(false, "");
+        prepareToCloseLocalTunnel(false, "");
         return true;
     }
 
 private:
+    friend class tests::TestMPPTunnel;
+
     bool checkPacketErr(TrackedMppDataPacketPtr & packet)
     {
         if (packet->hasError())
         {
-            closeLocalTunnel(true, packet->error());
+            prepareToCloseLocalTunnel(true, packet->error());
```

Comment:
How about rename `prepareToCloseLocalTunnel` to `finishWrite`, which means the local tunnel will guarantee that no more call to `local_request_handler.write` after `finishWrite` is called?

Reply:
> How about rename `prepareToCloseLocalTunnel` to `finishWrite`, which means the local tunnel will guarantee that no more call to `local_request_handler.write` after `finishWrite` is called?

okk